File: Leetcode/Python3/147. Insertion Sort List/solution.py
# https://leetcode.com/problems/insertion-sort-list/


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:  # My first answer
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)

        def insert(node):
            if not dummy.next:
                dummy.next = node
                return

            prev, cur = dummy, dummy.next
            while cur:
                if cur.val > node.val:
                    prev.next = node
                    node.next = cur
                    return
                prev, cur = cur, cur.next

            prev.next = node
            node.next = None
        pre = None
        n = head
        while n:
            pre = n
            n = n.next
            pre.next = None
            insert(pre)
        return dummy.next


class Solution:  # Second answer
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cur = dummy = ListNode(0)
        while head:
            while cur.next and cur.next.val < head.val:
                cur = cur.next

            cur.next, head.next, head = head, cur.next, head.next
            cur = dummy
        return dummy.next


class Solution:  # Third answer
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cur = dummy = ListNode(0)
        while head:
            while cur.next and cur.next.val < head.val:
                cur = cur.next

            cur.next, head.next, head = head, cur.next, head.next
            if head and cur.val > head.val:
                cur = dummy
        return dummy.next

File: Leetcode/C++/147. Insertion Sort List/solution.cpp
// https://leetcode.com/problems/insertion-sort-list/

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *insertionSortList(ListNode *head)
    {
        ListNode *root = new ListNode();
        ListNode *cur = root, *temp;
        // head = root of not sorted listnodes
        while (head)
        {
            while (cur->next && cur->next->val < head->val) // Move to insert index
            {
                cur = cur->next;
            }
            temp = cur->next;
            cur->next = head;
            head = head->next;
            cur->next->next = temp;
            // Id head->val smaller than cur->val head cant come after cur index so return to root of sorted listnodes
            if (head && cur->val > head->val)
            {
                cur = root;
            }
        }
        return root->next;
    }
};